SJRK-41: opening pull request for review to merge to organizational repo

BlueSlug opened the pull request:
Effectively a duplicate of PR from waharnum: https://github.com/fluid-project/fluid-couch-config/pull/1
Includes work recommended by reviewer

waharnum committed:
Save before Infusion attempt.

waharnum committed:
Infusionizing.

waharnum committed:
Work when view does not exist.

waharnum committed:
Don't update if views are unchanged.

waharnum committed:
Configurable design document name.

waharnum committed:
Function names for views can now be strings.

waharnum committed:
Use fluid.getGlobalValue for retrieving function names by strings.

waharnum committed:
Use expander in invoker.

waharnum committed:
Extract foundation grade.

waharnum committed:
Separate grades, add document insertion.

waharnum committed:
Separate autoconfig grade.

waharnum committed:
Update README.

waharnum committed:
Restructuring.

waharnum committed:
Documentation and cleanup.

waharnum committed:
README and comments.

waharnum committed:
Linting.

waharnum committed:
Refactoring, adding capacity to include a validation function.

waharnum committed:
Fully implement validation.

waharnum committed:
Linting.

waharnum committed:
Separating into different config grades.

waharnum committed:
Small comment updates.

BlueSlug committed:
NOJIRA: started to add testing

BlueSlug committed:
Merge remote-tracking branch 'upstream/master'

BlueSlug committed:
NOJIRA: fixed testing to use gpii-pouchdb and work from node

waharnum committed:
Merge pull request #1 from BlueSlug/master

Added initial testing and LICENCE.txt

waharnum committed:
Add missing gpii-pouchdb dev dependency, move kettle to devDependencies.

BlueSlug committed:
NOJIRA: added test for updateDocuments

BlueSlug committed:
linting...

waharnum committed:
Add example linter directives to couchConfigBasics.js to stop linter complaining about Couch global/args.

waharnum committed:
Linting.

waharnum committed:
Restructuring.

waharnum committed:
Simplify design document function, extend example usage.

waharnum committed:
Update example.

BlueSlug committed:
NOJIRA: merging from upstream/master

BlueSlug committed:
Merge remote-tracking branch 'upstream/master'

BlueSlug committed:
SJRK-6: added test for updateDesignDoc

waharnum committed:
Merge pull request #2 from BlueSlug/SJRK-6

SJRK-6: add tests using gpii-pouchdb

waharnum committed:
SJRK-6: fire the onDocsUpdated event only when all documents are processed.

BlueSlug committed:
NOJIRA: updated infusion to latest dev release

waharnum committed:
Merge pull request #3 from BlueSlug/SJRK-6

NOJIRA: updated infusion to latest dev release

waharnum committed:
SJRK-41: update license files & repo URLs; prepare for merge PR.

BlueSlug committed:
SJRK-41: implementing changes recommended by Simon for previous pull request. Detail in comments here: https://github.com/fluid-project/fluid-couch-config/pull/1

BlueSlug committed:
SJRK-41: replacing references to 'sjrk' with 'fluid'

simonbates commented:
`grunt lint` is failing for me:

```
Running "eslint:all" (eslint) task

fluid-couch-config/src/couchConfig.js
  118:55  error  Infix operators must be spaced  space-infix-ops

fluid-couch-config/tests/js/couchConfigTests.js
  143:80  error  'userCtx' is defined but never used     no-unused-vars
  152:9   error  'emit' is not defined                   no-undef
  157:76  error  'rerereduce' is defined but never used  no-unused-vars
  158:12  error  'sum' is not defined                    no-undef

✖ 5 problems (5 errors, 0 warnings)

Warning: Task "eslint:all" failed. Use --force to continue.

Aborted due to warnings.
```

BlueSlug committed:
SJRK-41: updates to pass linting, fixed typo in testReduceFunction

BlueSlug commented:
It should pass `grunt lint` now. I fixed a typo ("rerereduce" was meant to be "rereduce") and added "userCtx" and "rereduce" to an ignore pattern for the no-unused-vars rule, same as with couchConfigBasic.js in the examples directory. For the no-undef rules, I added those two variables to the global declaration, since those are functions in CouchDB's View syntax.

simonbates commented on examples/js/couchConfigBasic.js:
```diff
@@ -0,0 +1,78 @@
+// such as views or validate_doc_update
+
+/* global emit, doc, newDoc, oldDoc, userCtx, secObj, sum, rereduce */
+/*eslint no-unused-vars: ["error", { "vars": "local", "argsIgnorePattern": "rereduce|doc|newDoc|oldDoc|userCtx|secObj" }]*/
```
I think that a more directed solution to the unused function parameters would be better here, rather than a file level modification to `no-unused-vars`. This solution will apply to all functions and also is dependent on the spelling of the parameters. I think a better solution would be to either remove the unused parameters from the function definitions, or, if we want to keep them for CouchDB common practise reasons, I would use a more focused exception such as:

    // eslint-disable-next-line no-unused-vars

On the line immediately above the function definition.

simonbates commented on tests/js/couchConfigTests.js:
```diff
@@ -0,0 +1,225 @@
+*/
+
+/* global fluid, emit, sum  */
+/*eslint no-unused-vars: ["error", { "vars": "local", "argsIgnorePattern": "rereduce|doc|newDoc|oldDoc|userCtx|secObj" }]*/
```
Same thoughts here as for `examples/js/couchConfigBasic.js` -- I think a more direct solution would be better.

simonbates commented on Gruntfile.js:
```diff
@@ -0,0 +1,28 @@
+// Declare dependencies
```
Please add a copyright header.

simonbates commented on Gruntfile.js:
```diff
@@ -0,0 +1,28 @@
+        // Project package file destination.
+        pkg: grunt.file.readJSON("package.json"),
+        eslint: {
+            all: ["src/**/*.js", "tests/**/*.js", "demos/**/*.js", "examples/**/*.js", "*.js"]
```
The `demos/**/*.js` entry is not needed.

simonbates commented on Gruntfile.js:
```diff
@@ -0,0 +1,28 @@
+            all: ["src/**/*.js", "tests/**/*.js", "demos/**/*.js", "examples/**/*.js", "*.js"]
+        },
+        jsonlint: {
+            all: ["package.json", ".jshintrc", "src/**/*.json", "tests/**/*.json", "demos/**/*.json", "!node_modules", "!src/lib/**", "!tests/lib/**"]
```
We don't need these entries: `"demos/**/*.json", "!node_modules", "!src/lib/**", "!tests/lib/**"`

simonbates commented on README.md:
```diff
@@ -1 +1,24 @@
+
+## Usage
+
+See `examples/couchConfigBasic.js` for an example of usage.
```
Should be `examples/js/couchConfigBasic.js` -- also maybe make a link?

simonbates commented on examples/js/couchConfigBasic.js:
```diff
@@ -0,0 +1,78 @@
+// These eslint directives prevent the linter from complaining about the use of
```
Please add a copyright header.

simonbates commented on index.js:
```diff
@@ -0,0 +1,9 @@
+"use strict";
```
Please add a copyright header.

BlueSlug committed:
SJRK-41: tidied up eslint directives and added copyright headers

simonbates commented on index.js:
```diff
@@ -0,0 +1,18 @@
+require("./src/couchConfig");
+
+// Register our content so that it can be referenced in other packages using `fluid.module.resolvePath("%gpii-binder/path/to/content")`
+fluid.module.register("couch-config", __dirname, require);
```
Let's update the module name to `"fluid-couch-config"`.

And it looks like the comment on line 17 was copied from gpii-binder.

BlueSlug replied:
For the comment, would it be appropriate simply to replace the reference to the gpii-binder module with this module's name?
i.e. `fluid.module.resolvePath("%fluid-couch-config/path/to/content")`

simonbates replied:
I would say either update as you suggest, or we could also just remove it.

simonbates commented on src/couchConfig.js:
```diff
@@ -0,0 +1,361 @@
+var isEqual = require("underscore").isEqual;
+var size = require("underscore").size;
+
+var sjrk = fluid.registerNamespace("sjrk");
```
Please update the namespace to `"fluid.couchConfig"`, replacing `"srjk.server.couchConfig"`.

simonbates commented on tests/js/couchConfigTests.js:
```diff
@@ -0,0 +1,225 @@
+
+var fluid = require("infusion");
+
+var sjrk  = fluid.registerNamespace("sjrk");
```
For the tests, let's use the namespace `"fluid.tests.couchConfig"`.

BlueSlug committed:
SJRK-41: updated module name and comment

BlueSlug committed:
SJRK-41: fluid.server.couchConfig --> fluid.couchConfig, tests likewise

BlueSlug committed:
SJRK-41: removed a couple of spare newlines to be consistent

BlueSlug committed:
Merge branch 'master' into SJRK-41

BlueSlug committed:
Merge pull request #1 from BlueSlug/SJRK-41

SJRK-41: re-namespacing project to fluid

BlueSlug committed:
Merge branch 'master' of https://github.com/BlueSlug/fluid-couch-config

simonbates commented:
I had a go at sketching a new structure for couchConfig.js:

https://gist.github.com/simonbates/f49ac89fad27dbb2c7441f70f4511bc8

The CouchDB configuration steps are now structured as separate components of grade [`fluid.couchConfig.action`](https://gist.github.com/simonbates/f49ac89fad27dbb2c7441f70f4511bc8#file-couchconfigsketch-js-L36-L41). The actions are modeled on [Kettle Middleware](https://github.com/fluid-project/kettle/blob/master/docs/Middleware.md), as components with an invoker called `doAction` that returns a promise.

Each action's `doAction` invoker is called with 2 arguments:

- `payload` -- a value that is passed from action to action (if actions need to communicate with one another, not used in my sketch)
- `couchOptions` -- the CouchDB configuration options

I have sketched actions for each of the steps:

- [createDbIfNotExist](https://gist.github.com/simonbates/f49ac89fad27dbb2c7441f70f4511bc8#file-couchconfigsketch-js-L43-L63)
- [updateDesignDocument](https://gist.github.com/simonbates/f49ac89fad27dbb2c7441f70f4511bc8#file-couchconfigsketch-js-L65-L92)
- [updateDocuments](https://gist.github.com/simonbates/f49ac89fad27dbb2c7441f70f4511bc8#file-couchconfigsketch-js-L94-L114)

The [`fluid.couchConfig`](https://gist.github.com/simonbates/f49ac89fad27dbb2c7441f70f4511bc8#file-couchconfigsketch-js-L5-L24) grade has a single invoker `configureCouch`, that initiates the configuration actions. Overall configuration success or error is communicated using `onSuccess` and `onError` events.

The actions are structured in a pipeline using the [`fluid.promise.fireTransformEvent()`](https://docs.fluidproject.org/infusion/development/PromisesAPI.html#fluidpromisefiretransformeventevent-payload-options) function.

The [`fluid.couchConfig.pipeline`](https://gist.github.com/simonbates/f49ac89fad27dbb2c7441f70f4511bc8#file-couchconfigsketch-js-L116-L153) grade extends `fluid.couchConfig` with the typical pipeline of: create DB, update design document, and update documents. (Maybe we can think of a better name than "pipeline", such as "typical"?)

At the bottom of the file I have an example CouchDB configuration: https://gist.github.com/simonbates/f49ac89fad27dbb2c7441f70f4511bc8#file-couchconfigsketch-js-L155-L165

My hope with this refactoring was to:

- Provide clear communication of success or error
    - The top level `fluid.couchConfig` grade has `onSuccess` and `onError` events
    - The configuration actions use promises to manage asynchronous operations and the promise API provided `resolve` and `reject` mechanism
- Maintain flexibility in composing configuration actions
    - Using `fluid.promise.fireTransformEvent()`
- Package configuration actions as separate components, to keep actions from interfering with one another
    - Actions may communicate, if needed, using the `payload` object that is passed from action to action

Some things that need a bit more thought:

- I'm using the `options` parameter of `fluid.promise.fireTransformEvent()` to pass the `couchOptions` -- but some keys here have special meaning (`reverse` and `filterNamespaces`) and it would be better to keep our CouchDB configuration data separate
    - We could accomplish isolation by introducing another layer: the `doAction` signature becomes `function (payload, options)`, where our data is at `options.couchOptions` or such
- Rather than using the `options` parameter, the CouchDB configuration options could be passed in the `payload` object -- this would require that actions reliably pass the data on, but would simplify the `doAction` signature to a single argument and would keep our data separate from the `fluid.promise.fireTransformEvent()` `options`
- `onSuccess` and `onError` are too general, `onCouchConfigured` and `onCouchConfigurationError` are probably better

waharnum commented:
Hey @simonbates, at first glance this looks like a very sensible refactoring structure to me. It's a more generalizable pattern for things needing to happen in a certain sequence and modeled (as you say) on how Kettle does things.

On the things that need more thought, I think the `options.couchOptions` isolation makes more sense to me. Passing the CouchDB configuration through the payload doesn't quite feel appropriate for me - to my mind, a payload should be fairly transient.

simonbates commented on src/couchConfig.js:
```diff
@@ -0,0 +1,357 @@
+                        console.log("DB " + dbName + " created");
+                        completionEvent.fire();
+                    } else {
+                        console.log("DB " + dbName + " could not be created");
```
Once we have a mechanism in place for reporting errors programmatically (through promises, events), we should be able to reduce the amount of `console.log()` calls. And any remaining logging that we want to keep in place for debugging, we should switch to `fluid.log()`.

BlueSlug replied:
Proper error logging via promises is now available, so the number of calls to `console.log()` has been reduced. When the code is switched to `fluid.log()`, though, no feedback is given when running the `couchConfigBasic` example setup. I.e. the various step-by-step updates are hidden, and only messages passed back to the `onError` listener are displayed.

waharnum replied:
By default, `fluid.log()` logs at the `INFO` logLevel priority, as described in the table at https://docs.fluidproject.org/infusion/development/CoreAPI.html#fluidloglevel.

This means plain `fluid.log()` messages won't be displayed unless a more verbose log level than the default is set - see the description of the logging behaviour at https://docs.fluidproject.org/infusion/development/CoreAPI.html#fluidsetloggingloglevel

Try one of the following and see if it results in the step-by-step update: 
* `fluid.log(message, fluid.logLevel.IMPORTANT)` when logging the message
* `fluid.setLogging(true)` before starting the configuration running

While I'm bad at remembering to use it myself, `fluid.log()` is useful because of the varying levels of configuration for logging messages based on priorities, but the default behaviour is one I've found confusing before. :)

BlueSlug replied:
Thanks for your help, @waharnum, using `fluid.setLogging(true)` has solved the issue, and the logging is much cleaner looking with the timestamps, now. I've made a commit which incorporates this change.

simonbates commented on tests/js/couchConfigTests.js:
```diff
@@ -0,0 +1,219 @@
+
+require("../../src/couchConfig");
+
+"use strict";
```
Double `"use strict"` -- also on line 12.

simonbates commented on tests/js/couchConfigTests.js:
```diff
@@ -0,0 +1,219 @@
+
+// A basic validation that checks the document to make sure its 'type' is 'test'
+// eslint-disable-next-line no-unused-vars
+fluid.tests.couchConfig.testValidateFunction = function (newDoc, oldDoc, userCtx) {
```
I think that we are using "test" in 2 different ways in our naming of functions here.

We are using it as a descriptor (as in, a function used in our testing):

- testValidateFunction
- testMapFunction
- testReduceFunction

And we are using it as a verb (as in, do the tests for this document):

- testDbDocument
- testDbView

I think that the first case is probably most consistent with our common usage: a test, a test case, a test environment. And for the verb case, we could use "assert", "verify", or "checkThat"?

simonbates commented on tests/js/couchConfigTests.js:
```diff
@@ -0,0 +1,219 @@
+    }
+});
+
+fluid.defaults("fluid.tests.couchConfig.couchConfigTester", {
```
I believe that we are missing some code branches in our testing:

- Update a document when it already exists and is the same (and verify that we didn't update it anyway)
- Update a document when it already exists and is different
- Update a design document when it already exists and is the same
- Update a design document when it already exists and is different
- Each of the different function types supported in `fluid.couchConfig.designDocument.generateViews()`

It would also be worthwhile testing when a database already exists with documents not in our configuration and verifying that they are left alone.

And once we have programmatic reporting of error conditions, testing error conditions.

BlueSlug replied:
To note: `fluid.couchConfig.designDocument.generateViews()` has been renamed and refactored into `fluid.couchConfig.updateDesignDocument.renderViewFunctions()` and `fluid.couchConfig.updateDesignDocument.renderFunctionString()`

BlueSlug commented:
I have a few notes with respect to error handling in the proposed refactoring:

- If an error is encountered, how do we handle it?
  - Of course, the `onError` event can have any listener attached to it, but it may help to do a few things "by default". As we discussed a little in person, there's the possibility of rolling back all updates to that point. The reasoning behind this action would be that the couch-config definition is intended to represent an initialized state for the database (though perhaps not!), and if this initialization isn't complete we wouldn't want to leave things in a half-baked state.
  - Perhaps this behaviour could be offered as an option in some way which would allow the implementing programmer to decide what the best course of action is.
  - An extending grade could be produced that implements this, perhaps called something like `fluid.couchConfig.atomic`
- In the case of `fluid.couchConfig.pipeline` failing at various stages:
  - `createDbIfNotExistAction`: if this step fails, then the other two should not be called. If the database can't be created or found, then it stands to reason we shouldn't attempt to run any document-related operations on it.
  - `updateDesignDocumentAction`: if this step fails, the next one should not be called. If our `validate_doc_update` function can't be guaranteed to be in place, then it stands to reason we shouldn't attempt to update or create any new documents.
  - `updateDocumentsAction`: this is addressed in my first point, above.
- What should be done with documents that are already in the database at the time couchConfig is called, but whose keys are not present in the couchConfig documents? It would make sense to leave them unchanged, though perhaps this can also be parametrised somehow and left to the programmer's discretion.

BlueSlug committed:
SJRK-41: remove duplicate "use strict" statement

simonbates commented:
Thanks @BlueSlug these are great thoughts.

I think that a reasonable first implementation would be for any error to stop the configuration process and for details of that error to be communicated to the code that initiated the configuration.

I think your suggestion to support rolling back changes is a very good one. I would be happy for that work to be done at a later stage and I think that this pull request can go in without it.

Regarding documents already in the database. I think a good starting point would be to leave any documents not covered in the configuration alone. And I think an option to clear out documents not in the configuration also makes sense as something that we could add later.

simonbates commented:
I realised that I didn't discuss propagation of error details in my sketch or comment above.

We can communicate errors by providing an object with error details to the `promise.reject()` call. Our pattern is to build an object like the following:

    promise.reject({
        isError: true,
        message: "SOME ERROR MESSAGE",
        // Other details relevant to the error, such as statusCode
    })

For examples, see:

- https://github.com/GPII/universal/blob/master/gpii/node_modules/journal/src/Journal.js#L94-L97
- https://github.com/fluid-project/kettle/blob/master/lib/dataSource-file.js#L87-L91

I've updated my sketch with error propagation and a sample action promise rejection:

- The `fluid.promise.fireTransformEvent()` resolve value and rejection error are propagated in the top level event
- I added a sample rejection call: https://gist.github.com/simonbates/f49ac89fad27dbb2c7441f70f4511bc8#file-couchconfigsketch-js-L85-L88
- Errors are logged: https://gist.github.com/simonbates/f49ac89fad27dbb2c7441f70f4511bc8#file-couchconfigsketch-js-L163

In addition, I set the `couchOptions` as a property on the `fluid.promise.fireTransformEvent()` `opttions`, as discussed with @waharnum above.

Details of the changes here: https://gist.github.com/simonbates/f49ac89fad27dbb2c7441f70f4511bc8/revisions#diff-0c00b126665afb7f30bc19d406dc04ba

BlueSlug committed:
SJRK-41: WIP, capturing refactoring work thus far

BlueSlug committed:
SJRK-41: WIP, continue refactoring and make code DRYer

BlueSlug committed:
SJRK-41: WIP, updated basic example, sort of working

BlueSlug committed:
SJRK-41: refactoring complete, working but with issues (race condition)

BlueSlug committed:
SJRK-41: Refactoring truly complete, tests worked and passing, bugfixes

BlueSlug committed:
SJRK-41: remove superfluous logging call

BlueSlug committed:
SJRK-41: renamed test functions

BlueSlug committed:
SJRK-41: preparing to add new tests, reversed order of doc/design tests

BlueSlug committed:
SJRK-41: adding new tests, test 5 currently failing

BlueSlug committed:
SJRK-41: added destroyDbIfExists action, fixed tests, all passing now

BlueSlug committed:
SJRK-41: added sequenceGrade elements, added new tests, refactored

BlueSlug committed:
SJRK-41: solved fluid.log issue with waharnum's suggestion

waharnum committed:
SJRK-41: nest views under a 'views' key in design document structure; update test.

waharnum committed:
SJRK-41: move dbDocuments and dbDesignDocuments key outside couchOptions block.

waharnum commented:
I was looking at the refactored code in the context of `sjrk-story-telling-server` using a CouchDB instance and noticed two things:

1) Per tests, it looks like the `dbDocuments` and `dbDesignDocuments` keys should be top-level in the grade options, not members of the `couchOptions` key as they are in the grade definition.

2) As part of the refactoring that created the `dbDesignDocuments` block and allowed for multiple design documents in a single configuration (a good change!), the need to structure views under a `views` key for them to work with CouchDB got lost. This results in a design document being pushed that doesn't have working views in the CouchDB context (and we don't have a test for this currently because all our testing takes place in PouchDB, which doesn't work in the same way...)

**Design Document Currently Produced (via Fauxton)**
![image](https://user-images.githubusercontent.com/5279051/33343141-4e16ec80-d452-11e7-8177-791b454a1b93.png)

**What Should Be Produced**
![image](https://user-images.githubusercontent.com/5279051/33343238-9edf4ec8-d452-11e7-9567-965b81360533.png)

I think the easiest way to fix this is by replicating the expected structure of a CouchDB design document in the grade configuration, so that all views must be nested under a `views` key, rather than at the same level as `validate_doc_update` - this also means we have an obvious extension point should we ever feel the need to support design document options other than `views` and `validate_doc_update`.

As part of figuring out what was going on, I made a fork that addresses the above two issues, which is here: https://github.com/waharnum/fluid-couch-config/tree/viewsFix

Sorry for the WALL OF TEXT, but I started updating `sjrk-story-telling-server`'s `dbSetup.js` file experimentally and ran into these issues, then started tracing them. :)

waharnum committed:
SJRK-41: update in-code comments.

waharnum committed:
SJRK-41: update the example document.

waharnum committed:
SJRK-41: Update example tag count view URL.

BlueSlug committed:
Merge pull request #2 from waharnum/viewsFix

SJRK-41: fix issues with structure of views in design documents

simonbates commented:
There are some linting errors:

```
$ grunt lint
Running "eslint:all" (eslint) task

fluid-couch-config/src/couchConfig.js
   21:21  error  Unexpected trailing comma           comma-dangle
  216:43  error  Missing space before opening brace  space-before-blocks

fluid-couch-config/tests/js/couchConfigTests.js
  418:66  error  Trailing spaces not allowed  no-trailing-spaces

✖ 3 problems (3 errors, 0 warnings)

Warning: Task "eslint:all" failed. Use --force to continue.

Aborted due to warnings.
```

simonbates commented on package.json:
```diff
@@ -0,0 +1,34 @@
+  },
+  "repository": {
+    "type": "git",
+    "url": "git+"
```
Missing URL

simonbates commented on package.json:
```diff
@@ -0,0 +1,34 @@
+  "author": "Fluid Project",
+  "license": "(BSD-3-Clause OR ECL-2.0)",
+  "bugs": {
+    "url": ""
```
Missing URL

simonbates commented on package.json:
```diff
@@ -0,0 +1,34 @@
+  "bugs": {
+    "url": ""
+  },
+  "homepage": "",
```
Do we have a URL that we could put here? Or maybe we just remove?

simonbates commented on .jshintrc:
```diff
@@ -0,0 +1 @@
+{"node": true}
```
Can we remove `.jshintrc`? It looks like we are using `eslint` for `grunt lint`.

If we remove, we can remove its entry in `Gruntfile.js`:

    jsonlint: {
        all: ["package.json", ".jshintrc", "src/**/*.json", "tests/**/*.json"]
    }

simonbates commented:
I'd like to have a chat about the inclusion of `package-lock.json`. I don't think we have included this in other projects yet?

For me, I'm still running NPM 3 and can't easily verify it.

simonbates commented on src/couchConfig.js:
```diff
@@ -0,0 +1,371 @@
+    dbDesignDocuments: {
+        // An object whose keys are the names of design documents to be
+        // created in the database. Each design document can have a
+        // collection of zero or more views under the key 'views', where th
```
Missing "e" at end of line: "where th**e**"

simonbates commented on src/couchConfig.js:
```diff
@@ -0,0 +1,371 @@
+"use strict";
+
+var fluid = require("infusion");
+fluid.setLogging(true);
```
I don't think that we should include this `fluid.setLogging(true)` here. I think it should be under control of the code that is using this file. Which in our case would only need to be:

- `examples/js/couchConfigBasic.js`

As I think the test-running machinery enables logging.

waharnum committed:
SJRK-64: porting over retrying configuration approach, give a basic test.

waharnum committed:
SJRK-64: move retrying functionality into a composition-based implementation.

waharnum committed:
SJRK-64: separate retrying code into own file.

waharnum committed:
SJRK-64: add test for retrying timeout.

waharnum committed:
SJRK-64: rename some grades.

waharnum committed:
SJRK-64: finished initial retrying implementation; restore larger set of tests from commenting out.

waharnum committed:
SJRK-64: update README.

BlueSlug committed:
Merge pull request #3 from waharnum/SJRK-64

SJRK-64: add retrying behaviour

BlueSlug committed:
Merge pull request #4 from BlueSlug/SJRK-41

SJRK-41 merge into master

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/fluid-project/fluid-couch-config?pullRequest=2) <br/>All committers have signed the CLA.

BlueSlug committed:
SJRK-77: added Code of Conduct file, de-linted

BlueSlug committed:
SJRK-77: added contribution guidelines

BlueSlug committed:
SJRK-77: removed ECL licence, updated copyright dates

BlueSlug committed:
NOJIRA: typo

BlueSlug committed:
NOJIRA: fixed issues with package.json and removed another ref to ECL

BlueSlug committed:
NOJIRA: removed further references to ECL

BlueSlug committed:
NOJIRA: added homepage URL

BlueSlug committed:
NOJIRA: removed jshint references

BlueSlug commented:
> I'd like to have a chat about the inclusion of `package-lock.json`. I don't think we have included this in other projects yet?
> 
> For me, I'm still running NPM 3 and can't easily verify it.

@simonbates I've found this article which indicates that it's meant to be included in source control, though it's safe to remove it and disable its presence entirely given the unexpected effects it can have:
https://medium.com/coinmonks/everything-you-wanted-to-know-about-package-lock-json-b81911aa8ab8

With this in mind, what do you think?

colinbdclark commented:
We have, in other contexts, chosen to omit lock files from source control. They're a source of noise in pull requests, and we have a policy of always pinning our dependencies to specific versions. I recommend we do so here as well.

BlueSlug committed:
SJRK-41: removed package-lock.json and added it to .gitignore list

BlueSlug commented:
@colinbdclark @simonbates I have removed the offending file. I agree with Colin's recommendation of pinning versions in the package file, thereby reducing the need for a package-lock file in the first place.

BlueSlug commented:
Ready for more review :)

This should probably be revisited when someone has the time, as we're using this code in a production environment now. Likewise for [fluid-grunt-couch-config](https://github.com/fluid-project/fluid-grunt-couch-config/pull/1)

amb26 commented on src/couchConfig.js:
```diff
@@ -0,0 +1,391 @@
+        dbName: null // To be provided
+    },
+    dbDocuments: {
+        // An object whose keys are the IDs of documents to
```
As noted in https://issues.fluidproject.org/browse/SJRK-318?focusedCommentId=40849 this usage is incompatible with the immutable options introduced in the FLUID-6145 branch. Given these documents don't seem to be read from anywhere it might be more appropriate to resolve them from returned promise values rather than mutating them in place in the arguments list.

BlueSlug committed:
SJRK-41: removed ECL, linted, updated dependencies and copyright dates

BlueSlug committed:
SJRK-41: further updating copyright messages

BlueSlug committed:
Merge branch 'SJRK-41'

BlueSlug committed:
SJRK-41: updated deps

BlueSlug commented:
As discussed with @amb26 in [the #fluid-work IRC channel on December 16, 2019](http://irc-logs.fluidproject.org/%23fluid-work/%23fluid-work.2019-12-16.log), `fluid-couch-config` does not currently violate immutable options, as the intended usage is for an integrator to set up their database documents and design documents declaratively as infusion components, and these keys are not otherwise written to.

With this in mind, ready for more review :)

amb26 commented on README.md:
```diff
@@ -1 +1,26 @@
+
+* Includes a behaviour-based grade (`fluid.couchConfig.retryingBehaviour`) to implement retrying functionality, which is useful when running initial configuration in distributed environments such as containerization
+
+## Usage
```
README needs to explain prerequisites for running tests

BlueSlug committed:
SJRK-41: updated README with steps to run tests

amb26 commented on package.json:
```diff
@@ -0,0 +1,34 @@
+  "devDependencies": {
+    "eslint-config-fluid": "1.1.0",
+    "fluid-grunt-eslint": "18.1.2",
+    "gpii-pouchdb": "1.0.15",
```
Could do with a general spring cleaning of all dependencies

amb26 replied:
Probably required since the version of leveldown referenced from this version doesn't build on node 12.x, at least on Windows
